### PR TITLE
[Google] Use plus-minus blocks for text_join_simple and controls_if blocks

### PR DESCRIPTION
**Pull request by mikeharv**
Two common blocks used in Sprite Lab have known bugs when using Google Blockly. On CDO Blockly, both blocks work, but the UI has been found to be unintuitive.

If/else blocks use a modal flyout UI for adding branching statements, shown here on CDO Blockly:|Text join blocks on CDO Blockly feature a different UI for adding additional inputs, not used anywhere else:
-|-
if|join

On Google Blockly, if/else blocks' modal flyout does not render correctly, although the blocks are at least labeled. |Text join blocks fail to create any fields or inputs at all, and all child blocks are lost:
-|-
image|image

A convenient solution exists to potentially solve for both problems, thanks to the `@blockly/block-plus-minus` plugin (demo here: https://google.github.io/blockly-samples/plugins/block-plus-minus/test/index.html)

Unfortunately, the plugin is not modular; importing it immediately attempts to install all of its included blocks, which we don't want. (Blockly issue filed here: https://github.com/google/blockly-samples/issues/2087)
We also need to customize the included mutators so they work in ways that are more consistent with our users' current experience.

## Proposal

Until or unless the Blockly team releases a new version with the flexibility we'd require, we do have the option of copying some of the source code directly to get our desired functionality. We can then modify the mutators to suit our needs and register them at a time that makes sense for our flow. These tasks make up the scope of this PR.

## Changes

1. Move `blocksCommon` logic for `text_join_simple` blocks out to Blockly version-specific files in `customBlocks/`.
This is a flow we have used to support other complex blocks, such as mini-toolbox blocks. Unlike those blocks, these are shared by other labs so they are placed in `commonBlocks.js` files. This is done for organization purposes, but also because we might want to delete `customBlocks/*/spritelabBlocks.js` files after the migration is complete.

2. Directly copy field and serialization helper files from @Blockly/blocks-plus-minus
Specifically, we need two new image fields (field_minus and field_plus), the serialization helpers, and the new block mutators. We can apply these new mutators when the Google Blockly Wrapper is first accessed. 

3. Further customize the new mutator for `text_join` and `text_join_simple` blocks
Not even mentioning the main UI change, two other key differences exist between CDO Blockly's `text_join_simple` mutator and the mutator sourced from the plus-minus plugin:
* Block text reads "join" and not "create text with"
* The minimum number of inputs is 2 and not 0.

The former change is a single-line change after importing our localizations. The latter change ended up involving a larger refactor to incorporate a new `MINIMUM_INPUTS` constant. I took this approach so that we could potentially supporting any number of minimum inputs in the future, if desired. Once merged, I will also share this approach with the Blockly team in case they are interested in working it into the plug.

4. Point our custom `text_join_simple` block definition to the `text_join` block definition from core Blockly, using our new custom mutator.
This effectively de-duplicates two similar blocks while supporting projects/levels that reference either type. I am assuming this is okay because the standard `text_join` is only used in a small number of levels (40), most of which are unpublished in curricula. The exceptions are for invisible blocks used in Play Lab levels in Course 4 (both of which have the potential of being deprecated), and it is assumed that changing the UI for these is fine. For reference, the current `text_join` mutator UI matches the original if/else mutator UI:
image

5. Customize the mutator for `controls_if` blocks.
For whatever reason, @Blockly/blocks-plus-minus treats `else` statements as a static configuration of the block. Both CDO Blockly and Core Google Blockly supporting users adding/removing the else statement as desired. I added this ability to the new custom mutator using the plus and minus fields. When a user presses `+` we will add an `else` statement if there isn't already one. `Else` statements will now also have a non-indexed `-` field to trigger their removal.

7. Patch our XML deserialization flow to move block attributes to the mutation.
All versions of `text_join` and `text_join_simple` use a stored integer to track the number of available inputs. CDO Blockly's XML serializer sets an `inputcount` attribute right on the block, while mainline Google Blockly expects an `items` value to be found in the mutator. The `inputcount` attribute was a custom feature added by CDO engineers: https://github.com/code-dot-org/blockly/commit/afb96ce83bd5a633bfde9527a251ec046d0f0c44
Since we were already recursively processing blocks to convert `can_disconnect_from_parent` to `movable`, I generalized this function to allow us to make any needed modifications to a single arbitrary block. I then added `addMutationToTextJoinBlocks` to make make the change described above. Fortunately, serialization for `controls_if` blocks was already working.

block |**CDO Blockly (baseline)** | **[blocks-plus-minus demo](https://google.github.io/blockly-samples/plugins/block-plus-minus/test/index.html)** | **Feature branch**
-|-|-|-
`controls_if`|if| ifelse| ifelseifelse
`text_join`, `text_join_simple`|join|createtextwith | joinplusminus

## Links

Jira:
- https://codedotorg.atlassian.net/browse/CT-123
- https://codedotorg.atlassian.net/browse/CT-206

## Testing story

I tested flipping a variety of projects from CDO to Google Blockly using any number of inputs in each block. I also confirmed that a block's editable state is still respected.

_Sample project as saved with CDO Blockly:_
image

_Sample project loaded with Google Blockly:_
image

## Follow-up work

If the Blockly team is interested in making the plugin more modular, we can potentially start to use it in favor of the copied fields and serialization helpers. We would still need to use our own custom mutators in order to keep our needed functionality.

When we are ready to start migrating labs that use functions with parameters, we can consider re-using these fields for that purpose. This is already supported directly in the plugin, shown below, but we'd likely need to customize it work with our other custom procedure block fields.
image

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/blockly/addons/cdoXml.js [hunk 1]
```diff
@@ -148,26 +148,76 @@ export function addMutationToBehaviorDefBlocks(blockElement) {
...
+    return;
+  }
+  const mutationElement =
+    blockElement.querySelector('mutation') ||
+    blockElement.ownerDocument.createElement('mutation');
+  // Place mutator before fields, values, and other nested blocks.
+  blockElement.insertBefore(mutationElement, blockElement.firstChild);
+
+  // We need to keep track of the expected number of inputs in order to create them all.
+  // If not, it needs a static behavior id in order to be translatable
```
Comment (molly-moen) [inline apps/src/blockly/addons/cdoXml.js]: Are the comments about behavior id and `usercreated` relevant here? I don't see any logic relating to those in this function.

File: apps/src/blockly/addons/cdoXml.js [hunk 2]
```diff
@@ -148,26 +148,76 @@ export function addMutationToBehaviorDefBlocks(blockElement) {
...
+  // expects this kind of extra state in a mutator.
+  const inputCount = blockElement.getAttribute('inputcount');
+  mutationElement.setAttribute('items', inputCount);
+}
+
+/**
+ * A helper function designed to process each individual block in an XML tree.
+ * @param {Element} block - The XML element for a single block.
+ */
+function processAllBlocks(block) {
```
Comment (molly-moen) [inline apps/src/blockly/addons/cdoXml.js]: the function title is a little confusing, since it processes a single block (with potential child blocks), but is called `processAllBlocks`. Maybe `processBlockAndChildren` or something? Or just clarify in the comment we are processing a block and its children.

File: apps/src/blockly/addons/plusMinusBlocks/if.js [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+  /**
+   * Returns the state of this block as a JSON serializable object.
+   * @returns {{elseIfCount: (number|undefined),
+   *     haseElse: (boolean|undefined)}} The state of this block, ie the else
+   *     if count and else state.
+   */
+  saveExtraState: function () {
+    if (!this.elseIfCount_ && !this.hasElse_) {
+      return null;
```
Comment (molly-moen) [inline apps/src/blockly/addons/plusMinusBlocks/if.js:72]: should we return an empty object instead of `null`?

File: apps/src/blockly/addons/plusMinusBlocks/if.js [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+    } else {
+      this.addElse_();
+    }
+  },
+
+  /**
+   * Callback for the minus field. Triggers "removing" the input at the specific
+   * index.
+   * @see removeInput_
+   * @param {number} index The index of the else-if input to "remove". Undefined for else input.
```
Comment (molly-moen) [inline apps/src/blockly/addons/plusMinusBlocks/if.js]: Based on comments below this `index` cannot be 0 right? Worth including that in the comment?

File: apps/src/blockly/addons/plusMinusBlocks/if.js [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+  /**
+   * Applies the given state to this block.
+   * @param {*} state The state to apply to this block, ie the else if count and
+   *     else state.
+   */
+  loadExtraState: function (state) {
+    const targetCount = state['elseIfCount'] || 0;
+    this.hasElse_ = state['hasElse'] || false;
+    if (this.hasElse_ && !this.getInput('ELSE')) {
+      this.appendStatementInput('ELSE')
```
Comment (molly-moen) [inline apps/src/blockly/addons/plusMinusBlocks/if.js]: can we use the `addElse_` helper method here and in `domToMutation` instead? (if this is out of scope because this is copied from blockly that's fine)

File: apps/src/blockly/addons/plusMinusBlocks/text_join.js [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+  /**
+   * Removes an input from the end of the block. If we are removing the last
+   * input this updates the block to have an 'EMPTY' top input.
+   * @this {GoogleBlockly.Block}
+   * @private
+   */
+  removePart_: function () {
+    this.itemCount_--;
+    this.removeInput('ADD' + this.itemCount_);
+    if (this.itemCount_ === 0) {
```
Comment (molly-moen) [inline apps/src/blockly/addons/plusMinusBlocks/text_join.js:143]: why can we go down to 0 items when the minimum inputs is 2?

File: apps/src/blockly/customBlocks/googleBlockly/commonBlocks.js [hunk 7]
```diff
@@ -0,0 +1,10 @@
+/**
+ * Defines blocks useful in multiple blockly apps
+ */
+
+export const blocks = {
+  installJoinBlock(blockly) {
+    blockly.Blocks.text_join_simple = blockly.Blocks.text_join;
```
Comment (molly-moen) [inline apps/src/blockly/customBlocks/googleBlockly/commonBlocks.js:9]: for my information, where are `blockly.Blocks.text_join` and `blockly.JavaScript.text_join` defined?

**Review by molly-moen — COMMENTED** [review]
Looks good, some clarifying comments/questions

File: apps/src/blockly/addons/cdoXml.js (hunk 2 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoXml.js]: It’s the block and it’s children. Thanks, I’ll clarify!

File: apps/src/blockly/addons/plusMinusBlocks/if.js (hunk 4 above)
Comment (mikeharv) [reply apps/src/blockly/addons/plusMinusBlocks/if.js]: Cannot be 0. Will document with a comment!

File: apps/src/blockly/addons/plusMinusBlocks/text_join.js (hunk 6 above)
Comment (mikeharv) [reply apps/src/blockly/addons/plusMinusBlocks/text_join.js:143]: We can’t, as it’s written, but there was no minimum in the original plugin. I thought it better to leave that functionality (but make in accessible) for now in case we wanted to support the curriculum in setting a minimum per-level or per-block. I also had the thought that the Blockly team might want to add that level of customization to the plugin if they update it. Happy to strip it down if those are bad/confusing reasons!

File: apps/src/blockly/customBlocks/googleBlockly/commonBlocks.js (hunk 7 above)
Comment (mikeharv) [reply apps/src/blockly/customBlocks/googleBlockly/commonBlocks.js:9]: Included with core Blockly

File: apps/src/blockly/addons/plusMinusBlocks/if.js (hunk 3 above)
Comment (mikeharv) [reply apps/src/blockly/addons/plusMinusBlocks/if.js:72]: Good question. This was the plugin’s return value and it does seem to work as is. I’ll investigate if there’s any difference as far as Blockly is concerned and consider what we do in other blocks as well.

File: apps/src/blockly/addons/plusMinusBlocks/if.js (hunk 5 above)
Comment (mikeharv) [reply apps/src/blockly/addons/plusMinusBlocks/if.js]: That would be a smaller refactor than some of the other customizations I made. I’ll take a look!

File: apps/src/blockly/addons/plusMinusBlocks/text_join.js (hunk 6 above)
Comment (molly-moen) [reply apps/src/blockly/addons/plusMinusBlocks/text_join.js:143]: I think it may be worth adding a comment explaining that, or just surrounding it with an `if` to protect from the case where we would go below our minimum item count.

File: apps/src/blockly/addons/plusMinusBlocks/if.js (hunk 3 above)
Comment (mikeharv) [reply apps/src/blockly/addons/plusMinusBlocks/if.js:72]: From https://developers.google.com/blockly/guides/create-custom-blocks/extensions#serialization_hooks:

> If your block is in its default state when it is serialized, then your `saveExtraState` method can return `null` to indicate this. If your `saveExtraState` method returns `null` then no `extraState` property is added to the JSON. This keeps your save file size small.

Going to leave this as is!

File: apps/src/blockly/addons/plusMinusBlocks/text_join.js (hunk 6 above)
Comment (mikeharv) [reply apps/src/blockly/addons/plusMinusBlocks/text_join.js:143]: That protection is already in place (below), but I will add a comment.

https://github.com/code-dot-org/code-dot-org/blob/85a9823b7ac45ba6b698b4efd343ca76e33e5f3b/apps/src/blockly/addons/plusMinusBlocks/text_join.js#L101-L107

File: apps/src/blockly/addons/cdoXml.js (hunk 1 above)
Comment (mikeharv) [reply apps/src/blockly/addons/cdoXml.js]: Not relevant.  They should have been deleted. Will remove!

**Review by molly-moen — APPROVED** [review]
Big improvement, good job!
